File: patterns/factory_method.py
# Factory Method example: document exporter.
from abc import ABC, abstractmethod
class Document(ABC):
    @abstractmethod
    def export(self) -> str: ...
class PDFDocument(Document):
    def export(self) -> str: return "PDF content"
class WordDocument(Document):
    def export(self) -> str: return "Word content"
class DocumentCreator(ABC):
    @abstractmethod
    def create_document(self) -> Document: ...  # Subclasses decide the concrete product.
    def deliver(self) -> str:
        doc = self.create_document()
        return f"Delivering {doc.export()}"
class PDFCreator(DocumentCreator):
    def create_document(self) -> Document: return PDFDocument()
class WordCreator(DocumentCreator):
    def create_document(self) -> Document: return WordDocument()
if __name__ == "__main__":
    print(PDFCreator().deliver())
    print(WordCreator().deliver())

File: patterns/factory_method.js
class Document {
  export() {
    throw new Error("Method export() must be implemented");
  }
}

class PDFDocument extends Document {
  export() {
    return "PDF content";
  }
}

class WordDocument extends Document {
  export() {
    return "Word content";
  }
}

class DocumentCreator {
  createDocument() {
    throw new Error("Method createDocument() must be implemented");
  }

  deliver() {
    const doc = this.createDocument();
    return `Delivering ${doc.export()}`;
  }
}

class PDFCreator extends DocumentCreator {
  createDocument() {
    return new PDFDocument();
  }
}

class WordCreator extends DocumentCreator {
  createDocument() {
    return new WordDocument();
  }
}

if (require.main === module) {
  console.log(new PDFCreator().deliver());
  console.log(new WordCreator().deliver());
}

module.exports = {
  Document,
  PDFDocument,
  WordDocument,
  DocumentCreator,
  PDFCreator,
  WordCreator,
};
